File: app/src/main/java/org/pocketworkstation/pckeyboard/TextEntryState.java
/*
 * Copyright (C) 2008 The Android Open Source Project
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package org.pocketworkstation.pckeyboard;

import android.content.Context;
import org.pocketworkstation.pckeyboard.Keyboard.Key;
import android.text.format.DateFormat;
import android.util.Log;

import java.io.FileOutputStream;
import java.io.IOException;
import java.util.Calendar;

public class TextEntryState {
    
    private static final boolean DBG = false;

    private static final String TAG = "TextEntryState";

    private static boolean LOGGING = false;

    private static int sBackspaceCount = 0;

    private static int sSessionCount = 0;

    private static int sTypedChars;

    private static int sActualChars;

    public enum State {
        UNKNOWN,
        START,
        IN_WORD,
        PUNCTUATION_AFTER_WORD;
    }

    private static State sState = State.UNKNOWN;

    private static FileOutputStream sKeyLocationFile;
    private static FileOutputStream sUserActionFile;
    
    public static void newSession(Context context) {
        sSessionCount++;
        sBackspaceCount = 0;
        sTypedChars = 0;
        sActualChars = 0;
        sState = State.START;

        if (LOGGING) {
            try {
                sKeyLocationFile = context.openFileOutput("key.txt", Context.MODE_APPEND);
                sUserActionFile = context.openFileOutput("action.txt", Context.MODE_APPEND);
            } catch (IOException ioe) {
                Log.e("TextEntryState", "Couldn't open file for output: " + ioe);
            }
        }
    }

    public static void endSession() {
        if (sKeyLocationFile == null) {
            return;
        }
        try {
            sKeyLocationFile.close();
            // Write to log file
            // Write timestamp, settings,
            String out = DateFormat.format("MM:dd hh:mm:ss", Calendar.getInstance().getTime())
                    .toString()
                    + " BS: " + sBackspaceCount
                    + " saved: " + ((float) (sActualChars - sTypedChars) / sActualChars)
                    + "\n";
            sUserActionFile.write(out.getBytes());
            sUserActionFile.close();
            sKeyLocationFile = null;
            sUserActionFile = null;
        } catch (IOException ioe) {

        }
    }
    
    public static void acceptedDefault(CharSequence typedWord, CharSequence actualWord) {
    }

    public static void backToAcceptedDefault(CharSequence typedWord) {
    }

    public static void manualTyped(CharSequence typedWord) {
        sState = State.START;
        displayState();
    }

    public static void acceptedTyped(CharSequence typedWord) {
    }

    public static void selectedForCorrection() {
    }

    public static void typedCharacter(char c, boolean isSeparator) {
        boolean isSpace = c == ' ';
        switch (sState) {
            case IN_WORD:
                if (isSpace || isSeparator) {
                    sState = State.START;
                }
                break;
            case START:
            case UNKNOWN:
            case PUNCTUATION_AFTER_WORD:
                if (!isSpace && !isSeparator) {
                    sState = State.IN_WORD;
                } else {
                    sState = State.START;
                }
                break;
        }
        displayState();
    }

    public static void backspace() {
        sBackspaceCount++;
        displayState();
    }

    public static void reset() {
        sState = State.START;
        displayState();
    }

    public static State getState() {
        if (DBG) {
            Log.d(TAG, "Returning state = " + sState);
        }
        return sState;
    }

    public static boolean isCorrecting() {
        return false;
    }

    public static void keyPressedAt(Key key, int x, int y) {
        if (LOGGING && sKeyLocationFile != null && key.codes[0] >= 32) {
            String out = 
                    "KEY: " + (char) key.codes[0] 
                    + " X: " + x 
                    + " Y: " + y
                    + " MX: " + (key.x + key.width / 2)
                    + " MY: " + (key.y + key.height / 2) 
                    + "\n";
            try {
                sKeyLocationFile.write(out.getBytes());
            } catch (IOException ioe) {
                // TODO: May run out of space
            }
        }
    }

    private static void displayState() {
        if (DBG) {
            //Log.w(TAG, "State = " + sState, new Throwable());
            Log.i(TAG, "State = " + sState);
        }
    }
}

File: app/src/main/java/org/pocketworkstation/pckeyboard/TextEntryState.kt
/*
 * Copyright (C) 2008 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package org.pocketworkstation.pckeyboard

import android.content.Context
import android.text.format.DateFormat
import android.util.Log
import org.pocketworkstation.pckeyboard.Keyboard.Key
import java.io.FileOutputStream
import java.io.IOException
import java.nio.charset.Charset
import java.util.Calendar

/**
 * Tracks text input state and provides debug logging support for keyboard interactions.
 *
 * This class implements a state machine to track the current text entry state, which is used
 * for autocorrection and diagnostic logging. The state transitions follow these rules:
 * - START: Initial state or after a separator/space. Transitions to IN_WORD when
 *   a non-separator character is typed.
 * - IN_WORD: Inside a word (non-space, non-separator characters). Transitions to
 *   START when a space or separator is typed.
 * - PUNCTUATION_AFTER_WORD: After punctuation following a word. Transitions based
 *   on next character type.
 * - UNKNOWN: Undefined state, treated similarly to START.
 *
 * The class provides optional debug logging to files (key.txt and action.txt) when
 * the LOGGING flag is enabled.
 */
class TextEntryState {

    enum class State {
        UNKNOWN,
        START,
        IN_WORD,
        PUNCTUATION_AFTER_WORD,
    }

    companion object {
        private const val DBG = false
        private const val TAG = "TextEntryState"

        private var LOGGING = false

        private var sBackspaceCount = 0
        private var sSessionCount = 0
        private var sTypedChars = 0
        private var sActualChars = 0

        private var sState: State = State.UNKNOWN

        private var sKeyLocationFile: FileOutputStream? = null
        private var sUserActionFile: FileOutputStream? = null

        /**
         * Initializes a new text entry session and resets state counters.
         *
         * @param context The Android application context, used to open log files if LOGGING is enabled
         */
        @JvmStatic
        fun newSession(context: Context) {
            sSessionCount++
            sBackspaceCount = 0
            sTypedChars = 0
            sActualChars = 0
            sState = State.START

            if (LOGGING) {
                try {
                    sKeyLocationFile = context.openFileOutput("key.txt", Context.MODE_APPEND)
                    sUserActionFile = context.openFileOutput("action.txt", Context.MODE_APPEND)
                } catch (ioe: IOException) {
                    Log.e(TAG, "Couldn't open file for output: $ioe")
                }
            }
        }

        /**
         * Closes the current text entry session and writes session statistics to the debug log.
         */
        @JvmStatic
        fun endSession() {
            if (sKeyLocationFile == null) return

            try {
                sKeyLocationFile!!.close()

                val out =
                    DateFormat.format("MM:dd hh:mm:ss", Calendar.getInstance().time).toString() +
                        " BS: " + sBackspaceCount +
                        " saved: " + ((sActualChars - sTypedChars).toFloat() / sActualChars) +
                        "\n"

                sUserActionFile!!.write(out.toByteArray(Charset.defaultCharset()))
                sUserActionFile!!.close()

                sKeyLocationFile = null
                sUserActionFile = null
            } catch (_: IOException) {
                // intentionally swallowed
            }
        }

        @JvmStatic
        fun acceptedDefault(typedWord: CharSequence, actualWord: CharSequence) {
            // no-op
        }

        @JvmStatic
        fun backToAcceptedDefault(typedWord: CharSequence) {
            // no-op
        }

        @JvmStatic
        fun manualTyped(typedWord: CharSequence) {
            sState = State.START
            displayState()
        }

        @JvmStatic
        fun acceptedTyped(typedWord: CharSequence) {
            // no-op
        }

        @JvmStatic
        fun selectedForCorrection() {
            // no-op
        }

        /**
         * Updates the text entry state based on the typed character.
         *
         * @param c The character that was typed
         * @param isSeparator true if the character is a word separator, false otherwise
         */
        @JvmStatic
        fun typedCharacter(c: Char, isSeparator: Boolean) {
            val isSpace = c == ' '
            when (sState) {
                State.IN_WORD -> {
                    if (isSpace || isSeparator) {
                        sState = State.START
                    }
                }
                State.START,
                State.UNKNOWN,
                State.PUNCTUATION_AFTER_WORD -> {
                    sState = if (!isSpace && !isSeparator) {
                        State.IN_WORD
                    } else {
                        State.START
                    }
                }
            }
            displayState()
        }

        /**
         * Records a backspace key press and updates debug logging.
         */
        @JvmStatic
        fun backspace() {
            sBackspaceCount++
            displayState()
        }

        @JvmStatic
        fun reset() {
            sState = State.START
            displayState()
        }

        /**
         * Returns the current text entry state.
         *
         * @return The current State of the text entry state machine
         */
        @JvmStatic
        fun getState(): State {
            if (DBG) {
                Log.d(TAG, "Returning state = $sState")
            }
            return sState
        }

        @JvmStatic
        fun isCorrecting(): Boolean = false

        /**
         * Logs the location and coordinates of a key press for debug analysis.
         *
         * @param key The Key object that was pressed
         * @param x The X coordinate of the touch event
         * @param y The Y coordinate of the touch event
         */
        @JvmStatic
        fun keyPressedAt(key: Key, x: Int, y: Int) {
            if (LOGGING && sKeyLocationFile != null && key.codes[0] >= 32) {
                val out =
                    "KEY: " + key.codes[0].toChar() +
                        " X: " + x +
                        " Y: " + y +
                        " MX: " + (key.x + key.width / 2) +
                        " MY: " + (key.y + key.height / 2) +
                        "\n"
                try {
                    sKeyLocationFile!!.write(out.toByteArray(Charset.defaultCharset()))
                } catch (_: IOException) {
                    // TODO: May run out of space
                }
            }
        }

        private fun displayState() {
            if (DBG) {
                Log.i(TAG, "State = $sState")
            }
        }
    }
}
